### ISSUE TITLE
bug(startuptime): cannot write startuptime to stdout like in vim

### ISSUE THREAD

**Issue by vegerot** [neovim/neovim#26787]
### Problem

In vanilla vim, I can write the startup to stdout by running

```sh
$ vim --startuptime /dev/stdout +qall
```

In Neovim this doesn't work.

### Steps to reproduce

```
$ # vanilla vim, not neovim
$ vim --clean --startuptime /dev/stdout +qall

times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.019  000.019: --- VIM STARTING ---
000.112  000.093: Allocated generic buffers
001.786  001.674: locale set
001.791  000.005: clipboard setup
001.793  000.002: window checked
002.754  000.961: inits 1
002.761  000.007: parsing arguments
002.763  000.002: expanding arguments
008.542  005.779: shell init
008.926  000.384: Termcap init
008.935  000.009: inits 2
009.009  000.074: init highlight
013.513  003.418  003.418: sourcing /usr/share/vim/vim90/filetype.vim
014.258  000.042  000.042: sourcing /usr/share/vim/vim90/ftplugin.vim
014.884  000.032  000.032: sourcing /usr/share/vim/vim90/indent.vim
018.561  000.555  000.555: sourcing /usr/share/vim/vim90/colors/lists/default.vim
018.727  001.329  000.774: sourcing /usr/share/vim/vim90/syntax/syncolor.vim
018.812  001.966  000.637: sourcing /usr/share/vim/vim90/syntax/synload.vim
018.882  003.273  001.307: sourcing /usr/share/vim/vim90/syntax/syntax.vim
018.916  009.491  002.726: sourcing $VIMRUNTIME/defaults.vim
018.930  000.430: sourcing vimrc file(s)
018.935  000.005: inits 3
018.937  000.002: reading viminfo
018.954  000.017: setting raw mode
```

```sh
$ nvim --clean --startuptime /dev/stdout +qall


times in msec
 clock   self+sourced   self:  sourced script
Assertion failed: (count && count <= buf->size), function rbuffer_consumed, file rbuffer.c, line 126.
                                                                                                     zsh: abort      nvim --clean --startuptime /dev/stdout +qall
                                                                                                                                                                 %
```

aaaaand my tty is insane now ;)

### Expected behavior

`nvim --startuptime /dev/stdout` should print the startup time to stdout.

### Neovim version (nvim -v)

NVIM v0.10.0-dev Build type: RelWithDebInfo LuaJIT 2.1.1703358377 

### Vim (not Nvim) behaves the same?

no, VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Nov 12 2023 01:16:04)

### Operating system/version

macOS 14.2

### Terminal name/version

tmux

### $TERM environment variable

screen-256color

### Installation

build from repo

**Comment by zeertzjq** [comment]
Can't reproduce on Linux or FreeBSD. Seems to be macOS-only issue.

**Comment by clason** [comment]
Can't reproduce on macOS, either. Seems to be a tmux-only issue.

**Comment by wookayin** [comment]
In my environment (macOS, arm64 M1 Pro), I can always reproduce this on the latest master inside and outside tmux.

```
nvim --clean --startuptime /dev/stdout +qall
```

> aaaaand my tty is insane now ;)

@vegerot `reset` to the rescue


Stacktrace:
```cpp
Process 75329 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #4: 0x000000010014761c nvim`rbuffer_consumed_compact + 220
nvim`rbuffer_consumed_compact:
->  0x10014761c <+220>: adrp   x0, 972
    0x100147620 <+224>: add    x0, x0, #0xb83            ; "rbuffer_consumed_compac
t"
    0x100147624 <+228>: adrp   x1, 989
    0x100147628 <+232>: add    x1, x1, #0xd41            ; "rbuffer.c"
Target 0: (nvim) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #0: 0x0000000187b9f11c libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x0000000187bd6cc0 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x0000000187ae6a50 libsystem_c.dylib`abort + 180
    frame #3: 0x0000000187ae5d6c libsystem_c.dylib`__assert_rtn + 284
  * frame #4: 0x000000010014761c nvim`rbuffer_consumed_compact + 220
    frame #5: 0x000000010020d8d0 nvim`receive_msgpack.llvm.7989737768189088788 + 34
4
    frame #6: 0x0000000100190320 nvim`read_cb.llvm.10104332701991089040 + 336
    frame #7: 0x00000001000ca268 nvim`uv__stream_io at stream.c:1143:7 [opt]
    frame #8: 0x00000001000ca13c nvim`uv__stream_io(loop=<unavailable>, w=0x0000000
13d81dab0, events=<unavailable>) at stream.c:1203:5 [opt]
    frame #9: 0x00000001000cf7b4 nvim`uv__io_poll(loop=0x00000001005860f8, timeout=
48) at kqueue.c:385:9 [opt]
    frame #10: 0x00000001000c1038 nvim`uv_run(loop=0x00000001005860f8, mode=UV_RUN_
ONCE) at core.c:447:5 [opt]
    frame #11: 0x00000001001c898c nvim`ui_client_run + 160
    frame #12: 0x0000000100319308 nvim`main + 17424
    frame #13: 0x000000018785d058 dyld`start + 2224
```

**Comment by glepnir** [comment]
seems like M mac only ? clason is intel mac if i remember correctly. can't reproduce on arch.

**Comment by clason** [comment]
No, it's Debug only. Duplicate of #23781

**Comment by zeertzjq** [comment]
OP is using RelWithDebInfo, which has assertions enabled.

**Comment by zeertzjq** [comment]
It's not a duplicate as it works fine on Linux with a Debug build.

**Comment by vegerot** [comment]
@clason is it debug only?  I build with `RelWithDebInfo` and am getting this error


```vim
NVIM v0.10.0-dev
Build type: RelWithDebInfo
LuaJIT 2.1.1703358377
Compilation: /Library/Developer/CommandLineTools/usr/bin/cc -O2 -g -Og -g -flto=thin -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow
-Wconversion -Wvla -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -fsigned-char -fstack-protector-strong -Wimplicit-fallthrough -fdi
agnostics-color=auto -Wl,-export_dynamic -DUNIT_TESTING -DHAVE_UNIBILIUM -D_GNU_SOURCE -DINCLUDE_GENERATED_DECLARATIONS -DNVIM_TS_HAS_SET_MAX_START_DEPTH -I/Users/m0c0j7y/
workspace/github.com/neovim/neovim/.deps/usr/include/luajit-2.1 -I/Users/m0c0j7y/workspace/github.com/neovim/neovim/.deps/usr/include -I/Users/m0c0j7y/workspace/github.com
/neovim/neovim/build/src/nvim/auto -I/Users/m0c0j7y/workspace/github.com/neovim/neovim/build/include -I/Users/m0c0j7y/workspace/github.com/neovim/neovim/build/cmake.config
 -I/Users/m0c0j7y/workspace/github.com/neovim/neovim/src -I/Library/Developer/CommandLineTools/SDKs/MacOSX14.2.sdk/usr/include -I/opt/homebrew/opt/gettext/include
```

**Comment by clason** [comment]
Any build that triggers asserts, yes (which `Release` doesn't).

**Comment by vegerot** [comment]
> Just a guess, but can anyone check if this change fixes the issue?
> 
> ```diff
> diff --git a/src/nvim/main.c b/src/nvim/main.c
> index f01b6ecc8..5f41919cb 100644
> --- a/src/nvim/main.c
> +++ b/src/nvim/main.c
> @@ -1482,7 +1482,7 @@ static void init_startuptime(mparm_T *paramp)
>  {
>    for (int i = 1; i < paramp->argc - 1; i++) {
>      if (STRICMP(paramp->argv[i], "--startuptime") == 0) {
> -      time_fd = fopen(paramp->argv[i + 1], "a");
> +      time_fd = os_fopen(paramp->argv[i + 1], "a");
>        time_start("--- NVIM STARTING ---");
>        break;
>      }
> ```

unfortunately this doesn't make a difference for me :(

**Comment by vegerot** [comment]
> Any build that triggers asserts, yes (which `Release` doesn't).

https://github.com/neovim/neovim/assets/5297725/2806307c-57be-4b19-82dd-43324d4294b4


Even Release builds don't work

**Comment by clason** [comment]
Works for me, is all I can say. What's your terminal?

**Comment by vegerot** [comment]
> Works for me, is all I can say. What's your terminal?

iTerm2.  I just tried in Terminal.app and it _kiiiiinda_ works, but very buggy.  

https://github.com/neovim/neovim/assets/5297725/9016f296-0e87-45f3-acf1-19579d483bae

**Comment by clason** [comment]
Which version? (Ignore terminal.app, it's broken.)

**Comment by vegerot** [comment]
> (Ignore terminal.app, it's broken.)

The output is less broken than iTerm's! 😁

> Which version?

- iTerm2: Build 3.4.23
- macOS: 14.2.1
- SoC: M1 Max
- neovim: Release build of 5b199513d
- zsh: zsh 5.9 (x86_64-apple-darwin23.0)

**Comment by clason** [comment]
Yeah, the releases are broken as well. Use the 3.5.0 betas.

**Comment by zeertzjq** [comment]
Actually I don't think it is supposed works in Vim either. In Vim the output ends at `settings raw mode`, not `--- VIM STARTED ---`.

**Comment by zeertzjq** [comment]
Closing as this is either not supported, a duplicate of #23781, or blocked by #25377.

**Comment by wookayin** [comment]
Another workaround: disable "buffering"

```bash
stdbuf -oL nvim --startuptime /dev/stdout +qall
```

This makes nvim not crash, whereas with buffering enabled (i.e. without `stdbuf -oL`) it always crashes in my environments. So I guess #25377 is not relevant, but this seems another instance of #23781.
